### ISSUE TITLE
Keep track of which account performs rule deletions

### ISSUE THREAD

**Issue by haozturk** [rucio/rucio#6360]
### Description

To the best of our knowledge, there is no way to tell who deleted a certain rule. The only information we know about a rule deletion is the time at which the rule is deleted. In CMS, we observe unusual rule deletions and we cannot do forensics due to lack of information. 

@dynamic-entropy @ericvaandering @panos512 fyi

### Motivation

To be able to debug rule deletions

### Change

At the moment, I don't have a concrete proposal on where this information should be added.

**Comment by bari12** [comment]
We discussed this a bit on the workshop and I think it is a good idea.

When `delete-rule` is called, it is probably relatively easy to just add another column to the `rules` table and add the account, issuing the deletion, there.

Where it is more complicated if the rule deletion is just a consequence of a did deletion, possibly in the future. E.g. a did is marked for expiration in the future. Now to trickle through this account information to the time when the did is actually deleted and consequently the rule is deleted is more complicated. We probably have to extend the columns on the did table as well to record the account information there. Not extensively complex, but it needs to be covered in multiple places so we do not only record a part of the deletions.

**Comment by dchristidis** [comment]
Yes, we can store who deleted both rules and DIDs. So if the column in the former is empty, one can assume that it was ‘Rucio itself’ who deleted the rule as a direct result of the latter.

But I am more curious as to how ‘delete’ is to be defined. What if a rule had a limited lifetime since its creation? I don’t see this as a blocker, merely something that we should try to investigate all possible workflows in advance.

One additional hurdle will be the communal use of group and service accounts.

**Comment by haozturk** [comment]
Thanks all for the replies. 

> When delete-rule is called, it is probably relatively easy to just add another column to the rules table and add the account, issuing the deletion, there.

Eventually, the information should be propagated to the `rules_history` table, right? I don't really know how `rules_history` table is populated, but if it's triggered by changes in the `rules` table and all the relevant information is propagated to the history table, then this makes sense to me.

> Where it is more complicated if the rule deletion is just a consequence of a did deletion, possibly in the future. E.g. a did is marked for expiration in the future. Now to trickle through this account information to the time when the did is actually deleted and consequently the rule is deleted is more complicated. We probably have to extend the columns on the did table as well to record the account information there. Not extensively complex, but it needs to be covered in multiple places so we do not only record a part of the deletions.

I agree that we should cover all cases which triggers rule deletions, not just deletions via rule clients. Keeping track of did deletions would be useful even if there is no rule associated with the did, because we can also have accounting for actions such as `rucio erase` even there is no rule for it.

> But I am more curious as to how ‘delete’ is to be defined. What if a rule had a limited lifetime since its creation? I don’t see this as a blocker, merely something that we should try to investigate all possible workflows in advance.

It'd be valuable to keep track of deletions by Rucio in case of rule expirations. 

> One additional hurdle will be the communal use of group and service accounts.

This is indeed a concern for CMS as well. Would it be possible to record the identity instead of the account?

**Comment by bari12** [comment]
> Eventually, the information should be propagated to the `rules_history` table, right? I don't really know how `rules_history` table is populated, but if it's triggered by changes in the `rules` table and all the relevant information is propagated to the history table, then this makes sense to me.

Yes, the history table is basically just a row-copy from the primary table, so it first needs to be there, the propagation is trivial then.

> > One additional hurdle will be the communal use of group and service accounts.
> 
> This is indeed a concern for CMS as well. Would it be possible to record the identity instead of the account?

We need to think a bit about this one, but the way rucio does authentication, this will be much harder.

**Comment by rodwalker** [comment]
I am interested in the reason a dataset is gone, in particular if it was part of the lifetime model. The dataset did is no longer in the Db, although some trace is there since I cannot re-use the did. I wonder if the did can be kept with some special final state.

The  rule history is there ...

Rule deletion
Account : ddmadmin
RSE expression : UKI-LT2-QMUL_DATADISK
Time : 2024-09-12 23:41:58

so I suppose this is a similar request to have more information here. One could imagine a "reason" field as well as DN. For the lifetime case, it would be enough for RUCIO_ACCOUNT=lifetime to have permission and be used. A limited list reasons would be more general though.

We might want to list all the datasets deleted by the lifetime model fitting some pattern between certain dates.

Cheers,
Rod.

**Comment by bari12** [comment]
Thank you for raising this. We have reviewed this issue as part of our regular issue review.

We see value in this request, but we do not have the capacity to work on it within the next twelve months. 

In line with our [issue management guideline](https://rucio.cern/documentation/developer/issue_management), we are labelling it `wishlist` and closing it. This does not mean it is forgotten: wishlist items are revisited regularly during component reviews and roadmap planning, and this issue can be re-opened and added back to the pipeline as soon as interest and capacity align.
